### PR TITLE
Fix extra lines generated when using clad::array or array_ref

### LINKED ISSUE #350: Fix extra lines generated when using clad::array.
Clad uses the [`isUnusedResult`](https://github.com/vgvassilev/clad/blob/edb192554fb3b45a5ebaa786227847779e16b4a0/lib/Differentiator/VisitorBase.cpp#L197) function to check for unneccessary statements and remove them from the generated code.

Since the array subscript expression of `clad::array` calls it's operator overload the function `isUnusedResult` is not able to determine that statement is not required.

```
double arr(3);
arr[1]; // <- isUnusedResult returns true for this statement

clad::array a(3);
a[1]; // <- isUnusedResult returns false for this statement but it should return true
```

To fix this `isUnusedResult` needs to be modified to return true for such statements.

### PR DESCRIPTION
**Pull request by vaithak**
fixes #350

### CODE REVIEW

File: include/clad/Differentiator/Array.h [hunk 1]
```diff
@@ -102,16 +102,19 @@ template <typename T> class array {
   /// Returns the size of the underlying array
   CUDA_HOST_DEVICE std::size_t size() const { return m_size; }
   /// Returns the ptr of the underlying array
-  CUDA_HOST_DEVICE T* ptr() const { return m_arr; }
-  CUDA_HOST_DEVICE T*& ptr_ref() { return m_arr; }
+  CUDA_HOST_DEVICE T* ptr() const __attribute__((pure)) { return m_arr; }
```
Comment (vgvassilev) [inline include/clad/Differentiator/Array.h:105]: We probably should put that in a macro in case we produce derivative codes that get compiled with msvc.

Comment (vaithak) [reply include/clad/Differentiator/Array.h]: Ah, I forgot that; added it now 👍🏼

**Review by parth-07 — APPROVED** [review]

File: include/clad/Differentiator/Array.h [hunk 2]
```diff
@@ -102,16 +109,16 @@ template <typename T> class array {
   /// Returns the size of the underlying array
   CUDA_HOST_DEVICE std::size_t size() const { return m_size; }
   /// Returns the ptr of the underlying array
-  CUDA_HOST_DEVICE T* ptr() const { return m_arr; }
-  CUDA_HOST_DEVICE T*& ptr_ref() { return m_arr; }
+  CUDA_HOST_DEVICE T* ptr() const PUREFUNC { return m_arr; }
```
Comment (vgvassilev) [inline include/clad/Differentiator/Array.h:112]:
```suggestion
  CUDA_HOST_DEVICE PUREFUNC T* ptr() const  { return m_arr; }
```

Can we move the attributes to go first?

Comment (vaithak) [reply include/clad/Differentiator/Array.h]: fixed 👍🏼

**Pull request merged**